srml/staking/src/lib.rs

```diff
@@ -925,20 +932,69 @@ impl<T: Trait> Module<T> {
 	///
 	/// Returns the new `SlotStake` value.
 	fn select_validators() -> BalanceOf<T> {
-		let maybe_elected_candidates = elect::<T, _, _, _>(
+		let maybe_elected_set = elect::<T, _, _, _>(
 			Self::validator_count() as usize,
 			Self::minimum_validator_count().max(1) as usize,
 			<Validators<T>>::enumerate(),
 			<Nominators<T>>::enumerate(),
 			Self::slashable_balance_of,
-			ElectionConfig::<BalanceOf<T>> {
-				equalize: false,
-				tolerance: <BalanceOf<T>>::sa(10 as u64),
-				iterations: 10,
-			}
 		);
 
-		if let Some(elected_candidates) = maybe_elected_candidates {
+		if let Some(elected_set) = maybe_elected_set {
+			let elected_stashes = elected_set.0;
+			let assignments = elected_set.1;
+
+			// helper closure.
+			let to_balance = |b: ExtendedBalance| <T::CurrencyToVote as Convert<ExtendedBalance, BalanceOf<T>>>::convert(b);
+			let to_votes = |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b) as ExtendedBalance;
+
+			// The return value of this is safe to be converted to u64.
+			// The original balance, `b` is within the scope of u64. It is just extended to u128
+			// to be properly multiplied by a ratio, which will lead to another value
+			// less than u64 for sure. The result can then be safely passed to `to_balance`.
+			// For now the backward convert is used. A simple `TryFrom<u64>` is also safe.
+			let ratio_of = |b, p| (p as ExtendedBalance) * to_votes(b) / ACCURACY;
```

Comment:
Should we use `checked_mul()` to be safe?

Reply:
The comment above exactly explains why this is actually safe. Will change to saturating just in case.

If the comment is not sufficient to explain why, I'd be happy to be further challenged.

`p` is a value between `0` and `ACCURACY`, while `b` is a normal `balance` type elevated to u128. Hence, `p / ACCURACY    * b` is indeed less than or equal to `b` and can be safely multiplied.

Reply:
But your line is `p * b / A`. I don't know how the CPU evaluates that; is it possible for `p * b` to overflow before division even occurs? Of course, if `p` and `b` are both u64 cast to u128, it's a total edge case that it comes close to overflow (u64::max * u64::max), but if there is no/little overhead for saturating then I don't see the harm.